From eea63fa97dd08c1546d1015ab5ac5cbb2ebb82f8 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@apache.org>
Date: Fri, 15 Jan 2016 16:52:53 +0000
Subject: [PATCH 1380/2848] MAPREDUCE-6554. MRAppMaster servicestart failing
 with NPE in MRAppMaster#parsePreviousJobHistory.
 Contributed by Bibin A Chundatt (cherry picked
 from commit
 9fbd579ab50f051c8ad63d316ca5a3e24822dcdd)

Change-Id: Ib5b9e7e436de0529a9c8f93e4f0d5081a02a12d5
---
 .../hadoop/mapreduce/jobhistory/EventReader.java   |   40 ++++++++++++--------
 1 file changed, 25 insertions(+), 15 deletions(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/jobhistory/EventReader.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/jobhistory/EventReader.java
index a68b775..0b25bdc 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/jobhistory/EventReader.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-core/src/main/java/org/apache/hadoop/mapreduce/jobhistory/EventReader.java
@@ -20,9 +20,16 @@
 
 import java.io.Closeable;
 import java.io.DataInputStream;
-import java.io.IOException;
 import java.io.EOFException;
+import java.io.IOException;
 
+import org.apache.avro.AvroRuntimeException;
+import org.apache.avro.Schema;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.specific.SpecificDatumReader;
 import org.apache.hadoop.classification.InterfaceAudience;
 import org.apache.hadoop.classification.InterfaceStability;
 import org.apache.hadoop.fs.FileSystem;
@@ -31,13 +38,6 @@
 import org.apache.hadoop.mapreduce.Counters;
 import org.apache.hadoop.util.StringInterner;
 
-import org.apache.avro.Schema;
-import org.apache.avro.io.Decoder;
-import org.apache.avro.io.DecoderFactory;
-import org.apache.avro.io.DatumReader;
-import org.apache.avro.specific.SpecificData;
-import org.apache.avro.specific.SpecificDatumReader;
-
 @InterfaceAudience.Private
 @InterfaceStability.Unstable
 public class EventReader implements Closeable {
@@ -68,14 +68,24 @@ public EventReader(DataInputStream in) throws IOException {
     this.version = in.readLine();
 
     Schema myschema = new SpecificData(Event.class.getClassLoader()).getSchema(Event.class);
-    this.schema = Schema.parse(in.readLine());
-    this.reader = new SpecificDatumReader(schema, myschema);
-    if (EventWriter.VERSION.equals(version)) {
-      this.decoder = DecoderFactory.get().jsonDecoder(schema, in);
-    } else if (EventWriter.VERSION_BINARY.equals(version)) {
-      this.decoder = DecoderFactory.get().binaryDecoder(in, null);
+    Schema.Parser parser = new Schema.Parser();
+    String eventschema = in.readLine();
+    if (null != eventschema) {
+      try {
+        this.schema = parser.parse(eventschema);
+        this.reader = new SpecificDatumReader(schema, myschema);
+        if (EventWriter.VERSION.equals(version)) {
+          this.decoder = DecoderFactory.get().jsonDecoder(schema, in);
+        } else if (EventWriter.VERSION_BINARY.equals(version)) {
+          this.decoder = DecoderFactory.get().binaryDecoder(in, null);
+        } else {
+          throw new IOException("Incompatible event log version: " + version);
+        }
+      } catch (AvroRuntimeException e) {
+        throw new IOException(e);
+      }
     } else {
-      throw new IOException("Incompatible event log version: " + version);
+      throw new IOException("Event schema string not parsed since its null");
     }
   }
   
-- 
1.7.9.5

